src/FakeItEasy/Core/FakeManager.cs

```diff
@@ -241,5 +267,22 @@ private void UpdateLastSequenceNumber(int sequenceNumber)
             while (sequenceNumber > last &&
                    sequenceNumber != Interlocked.CompareExchange(ref this.lastSequenceNumber, sequenceNumber, last));
         }
+
+        /// <summary>
+        /// A hack to allow shared object call rules. Some of these rules need access to a FakeManager when executing, so
+        /// it'll be supplied in the <see cref="ApplyBestRule(IInterceptedFakeObjectCall, out IFakeObjectCallRule)"/> method.
+        /// The only reason these rules extend <see cref="IFakeObjectCallRule"/> is that they are emitted through
+        /// <see cref="IInterceptionListener.OnAfterCallIntercepted"/>.
+        /// </summary>

```

Comment:
As I hint at, this hack wouldn't be necessary if `IFakeObjectCallRule` didn't expose `Apply`. But it needs to in order for people to create their own rules.

But I don't think that `IInterceptionListener.OnAfterCallIntercepted` needs to be passed an object that has an `Apply` method; there's no need for `OnAfterCallIntercepted` to call `Apply`.
I don't even know why it would need to call `IsApplicable` or `NumberOfTimesToCall`. Shall we explore  changing this in a future major version? We could supply `OnAfterCallIntercepted` with a different interface that's a restricted version of `IFakeObjectCallRule`, or even remove the `IFakeObjectCallRule` parameter.

Reply:
Seems like a good idea. I don't see any value in having the call rule available in `OnAfterCallIntercepted`, except for information purposes. Maybe we could just pass a string describing the rule, or just the type of the rule.